File: 0018_4sum/main.cpp
#include <vector>
#include <algorithm>

class Solution {
private:
    void kSum(std::vector<std::vector<int>>& res,std::vector<int>& nums, std::vector<int>& quad, int k, int start, long long target) {
        if(nums.size() < k)
            return;
        if(k != 2) {
            for(int i = start; i < nums.size()-k+1; ++i) {
                if(i > start && nums[i] == nums[i-1])
                    continue;
                
                quad.push_back(nums[i]);
                kSum(res, nums, quad, k-1, i+1, target-nums[i]);
                quad.pop_back();
            }
            return;
        }

        int l = start;
        int r = nums.size()-1;
        long long sum;
        while(l < r) {
            sum = nums[l] + nums[r];
            if(sum < target)
                ++l;
            else if(sum > target)
                --r;
            else {
                quad.push_back(nums[l]);
                quad.push_back(nums[r]);
                res.push_back(quad);
                quad.pop_back();
                quad.pop_back();
                ++l;
                while(l < r && nums[l] == nums[l-1])
                    ++l;
            }
        }
        return;
    }
public:
    std::vector<std::vector<int>> fourSum(std::vector<int>& nums, int target) {
        std::vector<std::vector<int>> res;
        std::vector<int> quad;
        std::sort(nums.begin(), nums.end());
        kSum(res, nums, quad, 4, 0, target);
        return res;
    }
};

int main(int argc, char** argv) {
    std::vector<int> test({1000000000,1000000000,1000000000,1000000000});
    Solution sol;
    sol.fourSum(test, -294967296);
}

File: 0018_4sum/main.py
from typing import List

class Solution:
    def fourSum(self, nums: List[int], target: int) -> List[List[int]]:
        nums.sort()
        res, quad = [], []

        def kSum(k, start, target):
            if k != 2:
                for i in range(start, len(nums) - k + 1):
                    if i > start and nums[i] == nums[i-1]:
                        continue
                    quad.append(nums[i])
                    kSum(k-1, i+1, target-nums[i])
                    quad.pop()
                return
            
            l, r = start, len(nums)-1
            while l < r:
                sum = nums[l] + nums[r]
                if sum < target:
                    l += 1
                elif sum > target:
                    r -= 1
                else:
                    res.append([*quad, nums[l], nums[r]])
                    # res.append(quad + [nums[l], nums[r]])
                    l += 1
                    while l < r and nums[l] == nums[l-1]:
                        l += 1

        kSum(4, 0, target)
        return res
    
Solution().fourSum([1,0,-1,0,-2,2], 0)